resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsLifecycleManager.scala

```diff
@@ -0,0 +1,146 @@
+          execIdsRemovedInThisRound += execId
+        case _ =>
+      }
+    }
+
+    // Reconcile the case where Spark claims to know about an executor but the corresponding pod
+    // is missing from the cluster. This would occur if we miss a deletion event and the pod
+    // transitions immediately from running io absent.
+    (schedulerBackend.getExecutorIds().map(_.toLong).toSet
+      -- snapshot.executorPods.keySet
+      -- execIdsRemovedInThisRound).foreach { missingExecutorId =>
+      if (removedExecutorsCache.getIfPresent(missingExecutorId) == null) {
+        val exitReason = ExecutorExited(
+          UNKNOWN_EXIT_CODE,
+          exitCausedByApp = false,
+          s"The executor with ID $missingExecutorId was not found in the cluster but we didn't" +
+            s" get a reason why. Marking the executor as failed. The executor may have been" +
+            s" deleted but the driver missed the deletion event.")
+        schedulerBackend.doRemoveExecutor(missingExecutorId.toString, exitReason)
+      }
+    }
+  }
+
+  private def removeExecutorFromK8s(updatedPod: Pod): Unit = {
+    // If deletion failed on a previous try, we can try again if resync informs us the pod
+    // is still around.
+    // Delete as best attempt - duplicate deletes will throw an exception but the end state
+    // of getting rid of the pod is what matters.
+    Utils.tryLogNonFatalError {
+      kubernetesClient
+        .pods()
+        .withName(updatedPod.getMetadata.getName)
+        .delete()
+    }
+  }
+
+  private def removeExecutorFromSpark(
+      schedulerBackend: KubernetesClusterSchedulerBackend,
+      pod: Pod,
+      execId: Long): Unit = {
+    if (removedExecutorsCache.getIfPresent(execId) == null) {
+      removedExecutorsCache.put(execId, execId)
+      val exitReason = findExitReason(pod, execId)
+      schedulerBackend.doRemoveExecutor(execId.toString, exitReason)
+    }
+  }
+
+  private def findExitReason(pod: Pod, execId: Long): ExecutorExited = {
+    val exitCode = findExitCode(pod)
+    val (exitCausedByApp, exitMessage) = if (isDeleted(pod)) {
+      (false, s"The executor with id $execId was deleted by a user or the framework.")
+    } else {
+      val msg = exitReasonMessage(pod, execId, exitCode)
+      (true, msg)
+    }
+    ExecutorExited(exitCode, exitCausedByApp, exitMessage)
+  }
+
+  private def exitReasonMessage(pod: Pod, execId: Long, exitCode: Int) = {
+    s"""
+       |The executor with id $execId exited with exit code $exitCode.
+       |The API gave the following brief reason: ${pod.getStatus.getReason}
+       |The API gave the following message: ${pod.getStatus.getMessage}
+       |The API gave the following container statuses:
+       |
+       |${pod.getStatus.getContainerStatuses.asScala.map(_.toString).mkString("\n===\n")}
+      """.stripMargin
+  }
+
+  private def isDeleted(pod: Pod): Boolean = pod.getMetadata.getDeletionTimestamp != null
```

Comment:
this method is duplicated in `ExecutorPodsSnapshot.scala`, can we get rid of it here?